### ISSUE TITLE
Block Commenting a Collapsed/Elided Chunk of Python Text Causes Line Numbering to Enter Faulty State

### ISSUE THREAD

**Issue by GaryBloom** [notepad-plus-plus/notepad-plus-plus#7796]
### Description of the Issue

While editing Python code, you can collapse/hide chunks of nested text (like in an If or Else statement) by clicking on the little red "minus" symbol to the right of the line number. Once collapsed, you can select the entire hidden chunk by selecting the first line of the hidden text and drag-selecting to the beginning of the next non-hidden line. Then, if you right-click on the first selected line and select "Block Comment", the first selected line and all the hidden lines become block commented, but the line numbering is now corrupt. The previously-hidden line numbers are gone from the line number panel and seem unrecoverable without reloading the file.

### Steps to Reproduce the Issue

1. create a simple four-line Python file as per the attached screenshot.
2. collapse the `if` statement
3. select the If statement and all hidden text by dragging to the beginning of the exposed `pass` statement
4. Right-click on the selected If statement and select 'Block Comment'

### Expected Behavior

I expect to see both of the `If` statement lines (lines 2 and 3) block-commented. 
### Actual Behavior

Line 2 is shown as block-commented, but line 3 (the hidden/elided text) is not shown, and line number 3 is gone from the line numbering panel, as well.
### Debug Information

![NPP Bug p1-3](https://user-images.githubusercontent.com/19311746/71561671-845db600-2a47-11ea-9e6f-0f6e38e9c983.jpg)

**Comment by An1onS** [comment]
I've checked this.
For me, it only happens if CR LF unselected
![изображение](https://user-images.githubusercontent.com/26001123/71621123-561ada80-2bef-11ea-9d04-f41866bb37e3.png)![изображение](https://user-images.githubusercontent.com/26001123/71620765-8eb9b480-2bed-11ea-9e79-a488c4f1e7fb.png)
Otherwise got expected behaviour
![изображение](https://user-images.githubusercontent.com/26001123/71621096-40a5b080-2bef-11ea-8e5f-8d54d64d2f60.png)![изображение](https://user-images.githubusercontent.com/26001123/71620808-bad53580-2bed-11ea-86e3-2b98bb4d412c.png)
Can you do so with 'Show All Characters' enabled?

**Comment by GaryBloom** [comment]
@An1onS  -  I'm running NP++ 7.8.2 x64 on Win 10. I see the same behavior whether CR-LF is selected or not.
If I select just one word on line 2 while line 3 is hidden, I get the same failure.
Also, there's a subtle error shown in your top two pics. Notice that before the Block Comment, the '+' is on line 2. But after the BC, a '-' appears at line 1. Two symptoms of the same problem.

**Comment by GaryBloom** [comment]
My quick thought  on logic for a fix (without looking at the code, sorry to say) is, before executing a Block Comment ("BC"), reveal any elided chunks of text related to the text selection.  I seem to recall this working better in the past, so this may be a regression.  Testing against some older versions may prove helpful.

**Comment by GaryBloom** [comment]
An alternative, perhaps less obvious but more useful, approach would be to support the eliding/collapsing of "nested" Python comments. So, a block that was collapsible before block commenting would still be collapsed after block commenting, because the "top" commented line would be more outdented than the lower lines.

**Comment by GaryBloom** [comment]
Turning that same elided block of text into a multi-line Python string (with triple quotes) also causes corruption of the line numbers.

**Comment by mchubby** [comment]
Looking at other reported issues, it resembles a slew of bugs related to the Scintilla component,
- language highlighting (built-in or UDL) folding/collapsing
- commenting/deleting/sub-comments

#4668 is similar to yours with curly brace-delimited scopes
#7836 for XML comments
#4560 collapsing code already containing comments, has interesting details

issues, some of them closed/marked as fixed, some saying it can't be fixed in npp/scintilla
#3910 #4133 #3213
#3638 #3962

"It's rather a bug in Scintilla" says the author in https://github.com/notepad-plus-plus/notepad-plus-plus/issues/7759#issuecomment-567856693_

deleting folded text:
#4585

indenting folded text:
#5856 #2013

To sum it up tidily:
"Folding is broken" #1162

**Comment by KnIfER** [comment]
It's not a bug in Scintilla. Notepad3, SciTE.exe, they all don't have this issue.

It's suprising that in NPP the folding could just corrupt on editing. No other editors behave this way.

**Comment by KnIfER** [comment]
fix: It's something about SC_AUTOMATICFOLD_CHANGE 

Instead of :
```
	// Let Scintilla deal with some of the folding functionality
	_mainEditView.execute(SCI_SETAUTOMATICFOLD, SC_AUTOMATICFOLD_SHOW);
	_subEditView.execute(SCI_SETAUTOMATICFOLD, SC_AUTOMATICFOLD_SHOW);
```

try:
```
	_mainEditView.execute(SCI_SETAUTOMATICFOLD, SC_AUTOMATICFOLD_SHOW|SC_AUTOMATICFOLD_CHANGE );
	_subEditView.execute(SCI_SETAUTOMATICFOLD, SC_AUTOMATICFOLD_SHOW|SC_AUTOMATICFOLD_CHANGE );
```

reference: https://www.scintilla.org/ScintillaDoc.html

**Comment by An1onS** [comment]
8.3.3 - fixed by #11280

**Comment by donho** [comment]
#11280
